tree.go

```diff
@@ -22,33 +22,12 @@ type Tree interface {
 	Root() ([]byte, error)
 }
 
-var _ Tree = &DefaultTree{}
+// trees is a global map used for keeping track of registered tree constructors for testing and JSON unmarshalling
+var trees = make(map[string]TreeConstructorFn)
 
-type DefaultTree struct {
-	*merkletree.Tree
-	leaves [][]byte
-	root   []byte
-}
-
-func NewDefaultTree(_ Axis, _ uint) Tree {
-	return &DefaultTree{
-		Tree:   merkletree.New(sha256.New()),
-		leaves: make([][]byte, 0, 128),
-	}
-}
-
-func (d *DefaultTree) Push(data []byte) error {
-	// ignore the idx, as this implementation doesn't need that info
-	d.leaves = append(d.leaves, data)
-	return nil
-}
-
-func (d *DefaultTree) Root() ([]byte, error) {
-	if d.root == nil {
-		for _, l := range d.leaves {
-			d.Tree.Push(l)
-		}
-		d.root = d.Tree.Root()
+func registerTree(treeName string, treeConstructor TreeConstructorFn) {
+	if trees[treeName] != nil {
+		panic(fmt.Sprintf("%s already registered", treeName))
```

Comment:
[question] instead of `panicking` can we modify the function signature to optionally return an error? If yes, then we can do something like this here:

```suggestion
		return fmt.Errorf("%s already registered", treeName)
```